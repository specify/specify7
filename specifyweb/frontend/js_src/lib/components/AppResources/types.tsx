import type { LocalizedString } from 'typesafe-i18n';

import { preferencesText } from '../../localization/preferences';
import { reportsText } from '../../localization/report';
import { resourcesText } from '../../localization/resources';
import type { IR, RA, RR } from '../../utils/types';
import { ensure } from '../../utils/types';
import { icons } from '../Atoms/Icons';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpAppResourceDir, Tables } from '../DataModel/types';
import type { AppResourceMode } from './helpers';

export type AppResourceScope =
  | 'collection'
  | 'discipline'
  | 'global'
  | 'user'
  | 'userType';

export type AppResourceType = {
  readonly tableName: keyof Tables & ('SpAppResource' | 'SpViewSetObj');
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
};

export const appResourceTypes: RR<AppResourceMode, AppResourceType> = {
  appResources: {
    tableName: 'SpAppResource',
    icon: icons.cog,
    label: resourcesText.appResource(),
  },
  viewSets: {
    tableName: 'SpViewSetObj',
    icon: icons.pencilAt,
    label: resourcesText.formDefinitions(),
  },
};

export type ScopedAppResourceDir = SerializedResource<SpAppResourceDir> & {
  readonly scope: AppResourceScope;
};

type AppResourceSubType = {
  readonly mimeType: string | undefined;
  readonly name: string | undefined;
  readonly documentationUrl: string | undefined;
  readonly icon: JSX.Element;
  readonly label: LocalizedString;
  /**
   * Whether when creating a new app resource of this type, should copy the
   * contents from an existing app resource of that type that is in current
   * scope.
   * Default value:
   * If app resource type can only have one specific name, this is true
   * Else false
   */
  readonly useTemplate?: boolean;
  /**
   * Only allow creating this app resource at certain levels
   */
  readonly scope?: RA<AppResourceScope>;
};

/**
 * The order of the subtypes matters. The filtering algorithm loops over these
 * in the order they are defined to find the first subType that matches the
 * current resource. Thus, subtypes should be sorted from the most
 * specific to the least specific.
 */
export const appResourceSubTypes = ensure<IR<AppResourceSubType>>()({
  label: {
    mimeType: 'jrxml/label',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.ticket,
    label: reportsText.label(),
  },
  report: {
    mimeType: 'jrxml/report',
    name: undefined,
    documentationUrl:
      'https://discourse.specifysoftware.org/t/creating-reports-labels-in-specify-7-jaspersoft-studio/628',
    icon: icons.documentReport,
    label: reportsText.report(),
  },
  userPreferences: {
    mimeType: 'application/json',
    name: 'UserPreferences',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/specify-7-user-preferences-webinar/861',
    icon: icons.cog,
    label: preferencesText.userPreferences(),
    useTemplate: false,
    scope: ['user'],
  },
  defaultUserPreferences: {
    mimeType: 'application/json',
    name: 'DefaultUserPreferences',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Setting-default-user-preferences',
    icon: icons.cog,
    label: preferencesText.defaultUserPreferences(),
  },
  collectionPreferences: {
    mimeType: 'application/json',
    name: 'CollectionPreferences',
    documentationUrl: undefined,
    icon: icons.cog,
    label: preferencesText.collectionPreferences(),
    scope: ['collection'],
  },
  leafletLayers: {
    mimeType: 'application/json',
    name: 'leaflet-layers',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Adding-Custom-Tile-Servers',
    icon: icons.locationMarker,
    label: resourcesText.leafletLayers(),
  },
  rssExportFeed: {
    mimeType: 'text/xml',
    name: 'ExportFeed',
    documentationUrl:
      'https://github.com/specify/specify7/wiki/Darwin-Core-Archive-Publishing',
    icon: icons.upload,
    label: resourcesText.rssExportFeed(),
  },
  expressSearchConfig: {
    mimeType: 'text/xml',
    name: 'ExpressSearchConfig',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/simple-search-config/183',
    icon: icons.search,
    label: resourcesText.expressSearchConfig(),
  },
  typeSearches: {
    mimeType: 'text/xml',
    name: 'TypeSearches',
    documentationUrl:
      'https://discourse.specifysoftware.org/t/adding-a-non-native-query-combo-box/859#h-1-type-search-definition-typesearch_defxml-8',
    icon: icons.documentSearch,
    label: resourcesText.typeSearches(),
  },
  webLinks: {
    mimeType: 'text/xml',
    name: 'WebLinks',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/weblinks.xml',
    icon: icons.externalLink,
    label: resourcesText.webLinks(),
  },
  uiFormatters: {
    mimeType: 'text/xml',
    name: 'UIFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/uiformatters.xml',
    icon: icons.hashtag,
    label: resourcesText.uiFormatters(),
  },
  dataObjectFormatters: {
    mimeType: 'text/xml',
    name: 'DataObjFormatters',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/backstop/dataobj_formatters.xml',
    icon: icons.variable,
    label: resourcesText.dataObjectFormatters(),
  },
  dataEntryTables: {
    mimeType: 'text/xml',
    name: 'DataEntryTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/fish/dataentry_task.xml',
    icon: icons.pencilAt,
    label: resourcesText.dataEntryTables(),
  },
  interactionsTables: {
    mimeType: 'text/xml',
    name: 'InteractionsTaskInit',
    documentationUrl:
      'https://github.com/specify/specify6/blob/master/config/common/interactionstask.xml',
    icon: icons.chat,
    label: resourcesText.interactionsTables(),
  },
  otherXmlResource: {
    mimeType: 'text/xml',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.code,
    label: resourcesText.otherXmlResource(),
  },
  otherJsonResource: {
    mimeType: 'application/json',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.adjustments,
    label: resourcesText.otherJsonResource(),
  },
  otherPropertiesResource: {
    mimeType: 'text/x-java-properties',
    name: undefined,
    documentationUrl: undefined,
    icon: icons.viewList,
    label: resourcesText.otherPropertiesResource(),
  },
  otherAppResources: {
    mimeType: undefined,
    name: undefined,
    documentationUrl: undefined,
    icon: icons.document,
    label: resourcesText.otherAppResource(),
  },
} as const);
