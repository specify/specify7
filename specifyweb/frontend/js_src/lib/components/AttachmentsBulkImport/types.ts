import type React from 'react';
import type { LocalizedString } from 'typesafe-i18n';
import type { State } from 'typesafe-reducer';

import type { RA, RR } from '../../utils/types';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { Attachment, Tables } from '../DataModel/types';
import type { DatasetBase, DatasetBriefBase } from '../WbPlanView/Wrapped';
import type { PartialAttachmentUploadSpec } from './Import';
import type { staticAttachmentImportPaths } from './importPaths';
import type { keyLocalizationMapAttachment } from './utils';

export type UploadAttachmentSpec = {
  readonly token: string;
  readonly attachmentLocation: string;
};

type Matched = State<
  'matched',
  {
    readonly id: number;
  }
>;

type Skipped = State<
  'cancelled' | 'skipped',
  {
    readonly reason: keyof typeof keyLocalizationMapAttachment;
  }
>;
type Success = State<
  'success',
  {
    readonly successType: 'deleted' | 'uploaded';
  }
>;
export type AttachmentStatus = Matched | Skipped | Success;

export type PartialUploadableFileSpec = Partial<UploadableFileSpec> &
  Pick<UploadableFileSpec, 'uploadFile'>;

type UploadableFileSpec = {
  readonly uploadFile: UnBoundFile;
  readonly matchedId: RA<number>;
  readonly status: AttachmentStatus;
  readonly disambiguated: number;
  readonly attachmentId: number;
  readonly uploadTokenSpec: UploadAttachmentSpec;
  /*
   * This is added because if an attachment was uploaded correctly, but a network error
   * happened while saving the resource, and we retry, we will loose the previous attachment
   */
  readonly attachmentFromPreviousTry?: SpecifyResource<Attachment>;
};

/*
 * Forcing keys to be primitive because objects would be
 * ignored during syncing to backend.
 */
export type BoundFile = Pick<File, 'name' | 'size' | 'type'>;

export type UnBoundFile = {
  readonly file: BoundFile | File;
  readonly parsedName?: string;
};

export type AttachmentWorkProgress = {
  readonly total: number;
  readonly uploaded: number;
  readonly type: 'interrupted' | 'safe' | 'stopped' | 'stopping';
  readonly retryingIn: number;
  readonly stoppedByUser: boolean;
};

export type AttachmentWorkRef = {
  // eslint-disable-next-line functional/prefer-readonly-type
  mappedFiles: RA<PartialUploadableFileSpec>;
  // eslint-disable-next-line functional/prefer-readonly-type
  uploadPromise: Promise<number | undefined>;
  readonly retrySpec: Record<number, number>;
};

export type AttachmentWorkStateProps = {
  readonly onCompletedWork: (
    uploadables: RA<PartialUploadableFileSpec> | undefined
  ) => void;
  readonly workProgress: AttachmentWorkProgress;
  readonly workRef: React.MutableRefObject<AttachmentWorkRef>;
  readonly onStop: () => void;
  readonly triggerNow: () => void;
  readonly dialogText: RR<
    'onAction' | 'onCancelled' | 'onCancelledDescription',
    LocalizedString
  >;
};

export type AttachmentDatasetBrief = DatasetBriefBase & {
  readonly uploaderstatus:
    | 'deleting'
    | 'deletingInterrupted'
    | 'main'
    | 'uploading'
    | 'uploadInterrupted'
    | 'validating';
};

export type AttachmentDataSetPlan = AttachmentDatasetBrief & {
  readonly uploadplan: {
    readonly staticPathKey: string;
  };
};

export type AttachmentDataSet = AttachmentDatasetBrief &
  DatasetBase & {
    readonly rows: RA<PartialUploadableFileSpec>;
    readonly uploadplan: PartialAttachmentUploadSpec;
    readonly uploadresult?: {
      readonly timestamp: string;
    };
  };

export type FetchedDataSet =
  | AttachmentDataSet &
      (
        | { readonly uploaderstatus: 'main' }
        | ({
            readonly uploaderstatus: 'deleting' | 'uploading';
            readonly rows: RA<PartialUploadableFileSpec>;
          } & {
            readonly uploadplan: {
              readonly staticPathKey: keyof typeof staticAttachmentImportPaths;
            };
          })
      );

export type WrappedActionProps<KEY extends keyof Tables> = {
  readonly uploadableFile: PartialUploadableFileSpec;
  readonly baseTableName: KEY;
  readonly dryRun: boolean;
  readonly triggerRetry?: () => void;
};
