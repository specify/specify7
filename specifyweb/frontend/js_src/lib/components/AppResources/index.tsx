import React from 'react';

import { useCachedState } from '../../hooks/useCachedState';
import { useErrorContext } from '../../hooks/useErrorContext';
import { resourcesText } from '../../localization/resources';
import type { GetOrSet } from '../../utils/types';
import { Container, H2 } from '../Atoms';
import { ProtectedTable, ProtectedTool } from '../Permissions/PermissionDenied';
import { SafeOutlet } from '../Router/RouterUtils';
import { AppResourcesAside } from './Aside';
import { AppResourcesFilters } from './Filters';
import type { AppResources } from './hooks';
import { useAppResources } from './hooks';

export function AppResourcesWrapper(): JSX.Element {
  return (
    <ProtectedTool action="read" tool="resources">
      <ProtectedTable action="read" tableName="Discipline">
        <ProtectedTable action="read" tableName="Collection">
          <ProtectedTable action="read" tableName="SpecifyUser">
            <AppResourcesDataFetcher />
          </ProtectedTable>
        </ProtectedTable>
      </ProtectedTable>
    </ProtectedTool>
  );
}

function AppResourcesDataFetcher(): JSX.Element | null {
  const getSetResources = useAppResources();
  const [resources] = getSetResources;
  useErrorContext('appResourcesData', resources);

  return typeof resources === 'object' ? (
    <AppResourcesView getSet={getSetResources as GetOrSet<AppResources>} />
  ) : null;
}

export type AppResourcesOutlet = { readonly getSet: GetOrSet<AppResources> };

function AppResourcesView({
  getSet,
}: {
  readonly getSet: GetOrSet<AppResources>;
}): JSX.Element {
  const [resources] = getSet;

  const conformations = useCachedState('appResources', 'conformation');
  const [filters] = useCachedState('appResources', 'filters');

  return (
    <Container.FullGray>
      <div className="flex flex-wrap items-center gap-4">
        <H2 className="text-2xl">{resourcesText.resources()}</H2>
        <div className="flex gap-2">
          <AppResourcesFilters initialResources={resources} />
        </div>
      </div>
      <div className="flex flex-1 flex-col gap-4 sm:h-0 sm:overflow-scroll md:flex-row">
        <AppResourcesAside
          conformations={conformations}
          filters={filters}
          isEmbedded={false}
          resources={resources}
        />
        <SafeOutlet<AppResourcesOutlet>
          getSet={getSet as GetOrSet<AppResources>}
        />
      </div>
    </Container.FullGray>
  );
}
