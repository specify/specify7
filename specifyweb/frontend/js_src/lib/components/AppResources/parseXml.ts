export function strictParseXml(xml: string): Element {
  const parsed = parseXml(xml);
  // eslint-disable-next-line functional/no-throw-statement
  if (typeof parsed === 'string') throw new Error(parsed);
  else return parsed;
}

export function parseXml(string: string): Element | string {
  const parsedXml = new globalThis.DOMParser().parseFromString(
    string,
    'text/xml'
  ).documentElement;

  // Chrome, Safari
  const parseError = parsedXml.getElementsByTagName('parsererror')[0];
  if (typeof parseError === 'object')
    return (parseError.children[1].textContent ?? parseError.innerHTML).trim();
  // Firefox
  else if (parsedXml.tagName === 'parsererror')
    return (
      parsedXml.childNodes[0].nodeValue ??
      parsedXml.textContent ??
      parsedXml.innerHTML
    ).trim();
  else return parsedXml;
}
