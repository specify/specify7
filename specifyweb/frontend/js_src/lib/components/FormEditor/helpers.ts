import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
