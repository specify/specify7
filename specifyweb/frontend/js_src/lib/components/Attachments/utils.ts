import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
