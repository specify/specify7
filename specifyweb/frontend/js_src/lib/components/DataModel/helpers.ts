import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
