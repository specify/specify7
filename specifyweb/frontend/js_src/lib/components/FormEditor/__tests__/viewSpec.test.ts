import { requireContext } from '../../../tests/helpers';
import { localized } from '../../../utils/types';
import { strictParseXml } from '../../AppResources/parseXml';
import { tables } from '../../DataModel/tables';
import { error } from '../../Errors/assert';
import { formatXmlForTests } from '../../Syncer/__tests__/utils';
import { syncers } from '../../Syncer/syncers';
import { toSimpleXmlNode, xmlToJson } from '../../Syncer/xmlToJson';
import { updateXml } from '../../Syncer/xmlToString';
import {
  exportsForTests,
  formDefinitionSpec,
  parseSpecifyProperties,
} from '../viewSpec';
import { testFormDefinition } from './testFormDefinition';

requireContext();

const { buildSpecifyProperties } = exportsForTests;

const cases = {
  '': {},
  'name=Agent;title=Catalog Agent': {
    name: 'Agent',
    title: 'Catalog Agent',
  },
  'name=PartialDateUI;df=catalogedDate;tp=catalogedDatePrecision': {
    name: 'PartialDateUI',
    df: 'catalogedDate',
    tp: 'catalogedDatePrecision',
  },
  'name=CollectingEvent;clonebtn=true': {
    name: 'CollectingEvent',
    clonebtn: 'true',
  },
  'align=left;': {
    align: 'left',
  },
  'align=right;fg=0,190,0': {
    align: 'right',
    fg: '0,190,0',
  },
  'name=LocalityGeoRef;title=Geo Ref;geoid=geography;locid=localityName;llid=5':
    {
      name: 'LocalityGeoRef',
      title: 'Geo Ref',
      geoid: 'geography',
      locid: 'localityName',
      llid: '5',
    },
  'a=b=c;d=e=f%3B;': {
    a: 'b=c',
    d: 'e=f;',
  },
  ';a=b;': {
    a: 'b',
  },
};

const inCases = {
  'a=;b': {
    a: '',
    b: '',
  },
  '=a;': {},
};

describe('parseSpecifyProperties', () =>
  void Object.entries({ ...cases, ...inCases }).forEach(([input, output]) =>
    test(`parses ${input}`, () =>
      expect(parseSpecifyProperties(input)).toEqual(output))
  ));

describe('buildSpecifyProperties', () =>
  void Object.entries(cases).forEach(([output, input]) =>
    test(`parses ${output}`, () => {
      const trimmed = output.endsWith(';') ? output.slice(0, -1) : output;
      expect(buildSpecifyProperties(input)).toEqual(
        trimmed.startsWith(';') ? trimmed.slice(1) : trimmed
      );
    })
  ));

test('Can edit form definition', () => {
  const xml = strictParseXml(`<viewdef>${testFormDefinition}</viewdef>`);
  const xmlNode = xmlToJson(xml);
  const simpleXmlNode = toSimpleXmlNode(xmlNode);
  const { serializer, deserializer } = syncers.object(
    formDefinitionSpec(tables.Accession)
  );
  const parsed = serializer(simpleXmlNode);
  const parsedRows = parsed.definitions[0];
  const updated = deserializer({
    ...parsed,
    columnDefinitions: [
      parsed.columnDefinitions[0],
      ...parsed.columnDefinitions,
    ],
    definitions: [
      {
        ...parsedRows,
        rows: [
          parsedRows.rows[0],
          [
            ...parsedRows.rows[0],
            {
              ...parsedRows.rows[0][1],
              definition:
                parsedRows.rows[0][1].definition.type === 'Label'
                  ? {
                      ...parsedRows.rows[0][1].definition,
                      label: localized('New Label'),
                    }
                  : error('Expected a label cell at this position'),
            },
          ],
          ...parsedRows.rows,
        ],
      },
    ],
  });
  const updatedXml = formatXmlForTests(updateXml(updated));
  expect(updatedXml).toMatchSnapshot();
});
