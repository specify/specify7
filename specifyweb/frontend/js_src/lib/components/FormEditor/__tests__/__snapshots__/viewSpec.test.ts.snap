// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can edit form definition 1`] = `
"<?xml version='1.0' encoding='UTF-8'?>
<viewdef>
  <desc>The Accession form.</desc>
  <enableRules/>
  <columnDef>100px,2px,175px,5px,140px,2px,161px,5px,75px,2px,160px,0px,15px,p:g</columnDef>
  <columnDef>100px,2px,175px,5px,140px,2px,161px,5px,75px,2px,160px,0px,15px,p:g</columnDef>
  <columnDef os='lnx'>115px,2px,185px,5px,160px,2px,180px,5px,105px,2px,156px,0px,15px,p:g</columnDef>
  <columnDef os='mac'>130px,2px,205px,5px,180px,2px,200px,5px,125px,2px,176px,0px,15px,p:g</columnDef>
  <columnDef os='exp'>p,2px,p:g,5px:g,p,2px,p:g,5px:g,p,2px,p:g,0px,p,p:g</columnDef>
  <rowDef auto='true' cell='p' sep='1px'/>
  <rows>
    <row>
      <cell type='field' id='1' name='accessionNumber' uitype='formattedtext' initialize=''/>
      <cell type='label' labelfor='2' initialize=''/>
      <cell type='field' id='2' name='status' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='3' initialize=''/>
      <cell type='field' id='3' name='type' uitype='combobox' initialize=''/>
    </row>
    <row>
      <cell type='field' id='1' name='accessionNumber' uitype='formattedtext' initialize=''/>
      <cell type='label' labelfor='2' initialize=''/>
      <cell type='field' id='2' name='status' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='3' initialize=''/>
      <cell type='field' id='3' name='type' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='2' label='New Label' initialize=''/>
    </row>
    <row>
      <cell type='field' id='1' name='accessionNumber' uitype='formattedtext' initialize=''/>
      <cell type='label' labelfor='2' initialize=''/>
      <cell type='field' id='2' name='status' uitype='combobox' initialize=''/>
      <cell type='label' labelfor='3' initialize=''/>
      <cell type='field' id='3' name='type' uitype='combobox' initialize=''/>
    </row>
    <row>
      <cell type='label' labelfor='12' initialize=''/>
      <cell type='field' id='12' name='dateAccessioned' uitype='formattedtext' uifieldformatter='Date' default='today' initialize=''/>
      <cell type='label' labelfor='6' initialize=''/>
      <cell type='field' id='6' name='dateReceived' uitype='formattedtext' uifieldformatter='Date' default='today' initialize=''/>
    </row>
  </rows>
</viewdef>"
`;
